controllers/clustercache/cluster_accessor.go

```diff
@@ -0,0 +1,492 @@
+// other reconcilers (e.g. the Cluster or Machine reconciler).
+func (ca *clusterAccessor) Connect(ctx context.Context) (retErr error) {
+	log := ctrl.LoggerFrom(ctx)
+
+	if ca.Connected(ctx) {
+		log.V(6).Info("Skipping connect, already connected")
+		return nil
+	}
+
+	log.Info("Connecting")
+
+	// Creating clients, cache etc. is intentionally done without a lock to avoid blocking other reconcilers.
+	connection, err := ca.createConnection(ctx)
+
+	ca.lock(ctx)
+	defer ca.unlock(ctx)
+
+	defer func() {
+		if retErr != nil {
+			log.Error(retErr, "Connect failed")
+			ca.lockedState.lastConnectionCreationErrorTimestamp = time.Now()
+		}
+	}()
+
+	if err != nil {
+		return err
+	}
+
+	log.Info("Connected")
+
+	// Only generate the clientCertificatePrivateKey once as there is no need to regenerate it after disconnect/connect.
+	// Note: This has to be done before setting connection, because otherwise this code wouldn't be re-entrant if the
+	// private key generation fails because we check Connected above.
+	if ca.lockedState.clientCertificatePrivateKey == nil {
+		log.V(6).Info("Generating client certificate private key")
+		clientCertificatePrivateKey, err := certs.NewPrivateKey()
+		if err != nil {
+			return errors.Wrapf(err, "error creating client certificate private key")
+		}
+		ca.lockedState.clientCertificatePrivateKey = clientCertificatePrivateKey
+	}
+
+	now := time.Now()
+	ca.lockedState.healthChecking = clusterAccessorLockedHealthCheckingState{
+		// A client was just created successfully, so let's set the last probe times.
+		lastProbeTimestamp:        now,
+		lastProbeSuccessTimestamp: now,
+		consecutiveFailures:       0,
+	}
+	ca.lockedState.connection = &clusterAccessorLockedConnectionState{
+		restConfig:   connection.RESTConfig,
+		restClient:   connection.RESTClient,
+		cachedClient: connection.CachedClient,
+		cache:        connection.Cache,
+		watches:      sets.Set[string]{},
+	}
+
+	return nil
+}
+
+// Disconnect disconnects a connection to the workload cluster.
+func (ca *clusterAccessor) Disconnect(ctx context.Context) {
+	log := ctrl.LoggerFrom(ctx)
+
+	if !ca.Connected(ctx) {
+		log.V(6).Info("Skipping disconnect, already disconnected")
+		return
+	}
+
+	ca.lock(ctx)
+	defer ca.unlock(ctx)
+
+	log.Info("Disconnecting")
+
+	// Stopping the cache is non-blocking, so it's okay to do it while holding the lock.
+	// Note: Stopping the cache will also trigger shutdown of all informers that have been added to the cache.
+	log.V(6).Info("Stopping cache")
+	ca.lockedState.connection.cache.Stop()
+
+	log.Info("Disconnected")
+
+	ca.lockedState.connection = nil
+}
+
+// HealthCheck will run a health probe against the cluster's apiserver (a "GET /" call).
+func (ca *clusterAccessor) HealthCheck(ctx context.Context) (bool, bool) {
+	log := ctrl.LoggerFrom(ctx)
+
+	ca.rLock(ctx)
+	restClient := ca.lockedState.connection.restClient
+	ca.rUnlock(ctx)
```

Comment:
Question: Do we have to protect from hitting a nil pointer here?

if `ca.lockedState.connection` is nil, we would have a panic. This panic would be recovered so the lock would still be held, right?

Note: ok for me if we **know** that this will never be the case.

Reply:
Based on the current flow we will only ever call this when we are connected. I added an additional check to make this more robust
```
	if !ca.Connected(ctx) {
		log.V(6).Info("Skipping health check, not connected")
		return false, false
	}
```